authentik/core/tasks.py

```diff
@@ -92,3 +95,56 @@ def clean_temporary_users(self: SystemTask):
             deleted_users += 1
     messages.append(f"Successfully deleted {deleted_users} users.")
     self.set_status(TaskStatus.SUCCESSFUL, *messages)
+
+
+@CELERY_APP.task(bind=True, base=SystemTask)
+@prefill_task
+def purge_password_history_table(self: SystemTask):
+    """Remove all entries from the core.models.UserPasswordHistory table"""
+    messages = []
+    try:
+        # n.b. a performance optimization to execute TRUNCATE
+        # instead of all().delete() would eliminate any FK checks.
+        UserPasswordHistory.objects.all().delete()
+    except Exception as err:
+        LOGGER.debug("Failed to purge core.models.UserPasswordHistory table.")
+        self.set_error(err)
+        return
+    messages.append("Successfully purged core.models.UserPasswordHistory")
+    self.set_status(TaskStatus.SUCCESSFUL, *messages)
+
+
+@CELERY_APP.task()
+def trim_user_password_history(user_pk: int):
+    """Removes rows from core.models.UserPasswordHistory older than
+    the `n` most recent entries.
+
+    The `n` is defined by the largest configured value for all bound
+    UniquePasswordPolicy policies.
+    """
+    unique_password_policies = UniquePasswordPolicy.objects.all()
+
+    # All enable policy bindings for UniquePasswordPolicy
+    enabled_bindings = PolicyBinding.objects.filter(policy__in=unique_password_policies).filter(
+        enabled=True
+    )
+
+    if not enabled_bindings.exists():
+        return
+
+    num_rows_to_preserve = 0
+    for binding in enabled_bindings:
+        if hasattr(binding.policy, "num_historical_passwords"):
+            num_rows_to_preserve = max(
+                num_rows_to_preserve, binding.policy.num_historical_passwords
+            )
+
+    preservable_row_ids = (
+        UserPasswordHistory.objects.filter(user__pk=user_pk)
+        .order_by("-created_at")[:num_rows_to_preserve]
+        .values_list("id", flat=True)
+    )
+    num_deleted, _ = UserPasswordHistory.objects.exclude(pk__in=list(preservable_row_ids)).delete()
```

Comment:
Those two queries can probably be merged into one

Reply:
Are you suggesting merging for performance or to avoid having 2 querysets assigned to two variables? 

imo this is not as readable:
```python

    num_deleted, _ = UserPasswordHistory.objects.exclude(
        pk__in=list(
            UserPasswordHistory.objects.filter(user__pk=user_pk)
            .order_by("-created_at")[:num_rows_to_preserve]
            .values_list("id", flat=True)
        )
    ).delete()
```

Reply:
Oh yeah, I missed the slicing. It should be fine as is then